python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -716,23 +868,107 @@ async def can_be_fulfilled(self, requirement: DataRequirement) -> Tuple[bool, Op
+            return True
+        else:
+            return False
+
+    def can_derive_realization_from_formulations(self, requirement: DataRequirement, job: Job) -> bool:
+        """
+        Test if possible to derive a satisfactory realization config dataset from the originating request.
+
+        Test whether it is possible to derive a realization config that will satisfy this requirement, using a
+        formulation configuration contain within the original request message for this job.
+
+        Because this deals specifically with NextGen realization config datasets, a few conditions will immediately
+        result in a return of ``False``:
+            - a requirement category value other than ``CONFIG``
+            - a requirement domain data format value other than ``NGEN_REALIZATION_CONFIG``
+            - an originating request message for the job that is not a ::class:`NGENRequest`
+
+        Parameters
+        ----------
+        requirement : DataRequirement
+            The requirement for which the capability to derive a realization config needs to be determined.
+        job : Job
+            The job having the given requirement.
+
+        Returns
+        -------
+        bool
+            Whether deriving an appropriate realization configuration is possible.
+        """
+        if requirement.category != DataCategory.CONFIG:
+            return False
+        elif requirement.domain.data_format != DataFormat.NGEN_REALIZATION_CONFIG:
+            return False
+
+        request = job.model_request
+        return (isinstance(request, AbstractNgenRequest) and request.is_intelligent_request
+                and isinstance(request.formulation_configs, PartialRealizationConfig))
+
+    async def derive_datasets(self, job: Job) -> List[DataRequirement]:
+        """
+        Derive any datasets as required for the given job awaiting its data.
+
+        Job is expected to be in the ``AWAITING_DATA`` status step.  If it is not, no datasets are derived, a warning is
+        logged, and an empty list is returned.
+
+        If in the right status, but initially any unfulfilled requirements of the job cannot have a dataset successfully
+        derived, a ::class:`DmodRuntimeError` is raised.
+
+        Parameters
+        ----------
+        job
+
+        Returns
+        -------
+        List[DataRequirement]
+            A list of the given job's data requirements for a which a fulfilling dataset was derived and associated.
+
+        Raises
+        -------
+        DmodRuntimeError
+            Raised if a job with the correct status has an initially unfulfilled requirement for which a satisfactory
+            dataset can not be derived by this function.
+        """
+        # Only do something if the job has the right status
+        if job.status_step != JobExecStep.AWAITING_DATA:
+            return []
+
+        results = []
+
+        for req in [r for r in job.data_requirements if r.fulfilled_by is None]:
```

Comment:
totally unnecessary, but you could use a generator comprehension here instead of a list comprehension :).

Reply:
I didn't consider this and am not terribly skilled with my use of generators, so the discussion below is meant for learning purposes, not justification of the previous approach.

Would that save on anything other than the memory needed for the list object and it's references (small considering the referenced objects already exist)?  I don't know off the top of my head if, behind the scenes, the way `for` loop implementations work get something out of a generator here instead, although it seems to be just lazy versus eager application of the filtering.

Reply:
> Would that save on anything other than the memory needed for the list object and it's references (small considering the referenced objects already exist)?


Yes, but first a little setup:

We have:

- `n` items where `n == len(job.data_requirements)`
- `k` filtered items where `r.fulfilled_by is None`

So in the case of using the list comprehension you would do `n` iterations that result in a new list of `k` length. Then loop over the new list `k` times. So the total runtime is `Θ(n+k)`. By using the generator, you don't have to do an extra `k` iterations. So, the runtime is `Θ(n)`.

Reply:
Ah, gotcha  Thanks for the explanation.  The PR has already been approved at this point, so I'm not going to change things now and mess that up, but I'll keep this in mind in the future.